File: educational-core/src/com/jetbrains/edu/learning/serialization/converter/json/ToSecondVersionJsonStepOptionsConverter.java
package com.jetbrains.edu.learning.serialization.converter.json;

import com.google.gson.*;
import com.google.gson.reflect.TypeToken;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.EditorFactory;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Type;
import java.util.List;

import static com.jetbrains.edu.learning.serialization.SerializationUtils.Json.*;
import static com.jetbrains.edu.learning.serialization.SerializationUtils.*;

public class ToSecondVersionJsonStepOptionsConverter implements JsonStepOptionsConverter {

  @NotNull
  @Override
  public JsonObject convert(@NotNull JsonObject stepOptionsJson) {
    Gson gson = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).create();
    final JsonArray files = stepOptionsJson.getAsJsonArray(FILES);
    if (files != null) {
      for (JsonElement taskFileElement : files) {
        JsonObject taskFileObject = taskFileElement.getAsJsonObject();
        JsonArray placeholders = taskFileObject.getAsJsonArray(PLACEHOLDERS);
        for (JsonElement placeholder : placeholders) {
          JsonObject placeholderObject = placeholder.getAsJsonObject();
          convertToAbsoluteOffset(taskFileObject, placeholderObject);
          convertMultipleHints(gson, placeholderObject);
          convertToSubtaskInfo(placeholderObject);
        }
      }
    }
    return stepOptionsJson;
  }

  private static void convertToAbsoluteOffset(@NotNull JsonObject taskFileObject, @NotNull JsonObject placeholderObject) {
    int line = placeholderObject.getAsJsonPrimitive(LINE).getAsInt();
    int start = placeholderObject.getAsJsonPrimitive(START).getAsInt();
    if (line == -1) {
      placeholderObject.addProperty(OFFSET, start);
    }
    else {
      Document document = EditorFactory.getInstance().createDocument(taskFileObject.getAsJsonPrimitive(TEXT).getAsString());
      placeholderObject.addProperty(OFFSET, document.getLineStartOffset(line) + start);
    }
  }

  private static void convertMultipleHints(@NotNull Gson gson, @NotNull JsonObject placeholderObject) {
    final String hintString = placeholderObject.getAsJsonPrimitive(HINT).getAsString();
    final JsonArray hintsArray = new JsonArray();

    try {
      final Type listType = new TypeToken<List<String>>() {
      }.getType();
      final List<String> hints = gson.fromJson(hintString, listType);
      if (hints != null && !hints.isEmpty()) {
        for (int i = 0; i < hints.size(); i++) {
          if (i == 0) {
            placeholderObject.addProperty(HINT, hints.get(0));
            continue;
          }
          hintsArray.add(hints.get(i));
        }
        placeholderObject.add(ADDITIONAL_HINTS, hintsArray);
      }
      else {
        placeholderObject.addProperty(HINT, "");
      }
    }
    catch (JsonParseException e) {
      hintsArray.add(hintString);
    }
  }

  private static void convertToSubtaskInfo(@NotNull JsonObject placeholderObject) {
    JsonArray subtaskInfos = new JsonArray();
    placeholderObject.add(SUBTASK_INFOS, subtaskInfos);
    JsonArray hintsArray = new JsonArray();
    hintsArray.add(placeholderObject.getAsJsonPrimitive(HINT).getAsString());
    JsonArray additionalHints = placeholderObject.getAsJsonArray(ADDITIONAL_HINTS);
    if (additionalHints != null) {
      hintsArray.addAll(additionalHints);
    }
    JsonObject subtaskInfo = new JsonObject();
    subtaskInfos.add(subtaskInfo);
    subtaskInfo.add(INDEX, new JsonPrimitive(0));
    subtaskInfo.add(HINTS, hintsArray);
    subtaskInfo.addProperty(POSSIBLE_ANSWER, placeholderObject.getAsJsonPrimitive(POSSIBLE_ANSWER).getAsString());
  }
}

File: educational-core/src/com/jetbrains/edu/learning/serialization/converter/json/ToSecondVersionJsonStepOptionsConverter.kt
package com.jetbrains.edu.learning.serialization.converter.json

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.JsonMappingException
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ObjectNode
import com.intellij.openapi.editor.EditorFactory
import com.jetbrains.edu.learning.serialization.SerializationUtils
import java.io.IOException

class ToSecondVersionJsonStepOptionsConverter : JsonStepOptionsConverter {
  override fun convert(stepOptionsJson: ObjectNode): ObjectNode {
    val files = stepOptionsJson[SerializationUtils.Json.FILES]
    if (files != null) {
      for (taskFileElement in files) {
        val placeholders = taskFileElement[SerializationUtils.Json.PLACEHOLDERS]
        for (placeholder in placeholders) {
          convertToAbsoluteOffset(taskFileElement, placeholder as ObjectNode)
          convertMultipleHints(placeholder)
          convertToSubtaskInfo(placeholder)
        }
      }
    }
    return stepOptionsJson
  }

  companion object {
    private fun convertToAbsoluteOffset(taskFileObject: JsonNode, placeholderObject: ObjectNode) {
      val line = placeholderObject[SerializationUtils.LINE].asInt()
      val start = placeholderObject[SerializationUtils.START].asInt()
      if (line == -1) {
        placeholderObject.put(SerializationUtils.OFFSET, start)
      }
      else {
        val document = EditorFactory.getInstance().createDocument(taskFileObject[SerializationUtils.Json.TEXT].asText())
        placeholderObject.put(SerializationUtils.OFFSET, document.getLineStartOffset(line) + start)
      }
    }

    private fun convertMultipleHints(placeholderObject: ObjectNode) {
      val hintString = placeholderObject[SerializationUtils.HINT].asText()
      val hintsArray = placeholderObject.putArray(SerializationUtils.ADDITIONAL_HINTS)
      try {
        val hints = ObjectMapper().readValue(hintString, object : TypeReference<List<String>>() {})
        if (hints != null && hints.isNotEmpty()) {
          for (i in hints.indices) {
            if (i == 0) {
              placeholderObject.put(SerializationUtils.HINT, hints[0])
              continue
            }
            hintsArray.add(hints[i])
          }
        }
        else {
          placeholderObject.put(SerializationUtils.HINT, "")
        }
      }
      catch (e: JsonMappingException) {
        hintsArray.add(hintString)
      }
      catch (e: IOException) {
        hintsArray.add(hintString)
      }
    }

    private fun convertToSubtaskInfo(placeholderObject: ObjectNode) {
      val subtaskInfo = ObjectMapper().createObjectNode()
      val subtaskInfos = placeholderObject.putArray(SerializationUtils.Json.SUBTASK_INFOS)
      val hintsArray = subtaskInfo.putArray(SerializationUtils.Json.HINTS)
      hintsArray.add(placeholderObject[SerializationUtils.HINT].asText())
      val additionalHints = placeholderObject[SerializationUtils.ADDITIONAL_HINTS]
      if (additionalHints != null) {
        for (hint in additionalHints) {
          hintsArray.add(hint)
        }
      }
      subtaskInfos.add(subtaskInfo)
      subtaskInfo.put(SerializationUtils.Json.INDEX, 0)
      subtaskInfo.put(SerializationUtils.Json.POSSIBLE_ANSWER, placeholderObject[SerializationUtils.Json.POSSIBLE_ANSWER].asText())
    }
  }
}
